### PR TITLE
UCT: Add implicit MR support

### PR DESCRIPTION
**Pull request by eladpe**
support odp, rcache and direct registration

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3964/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3965/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3966/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3979/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3982/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3983/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3987/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3988/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by eladpe** [comment]
BOT:ornl:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3999/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4004/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
bot:MellanoxLab:retest

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by eladpe** [comment]
bot:mlxlab:retest

**Comment by eladpe** [comment]
bot:mlx:retest

**Comment by eladpe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4010/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4016/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
@yosefe ready for review

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -1255,6 +1305,12 @@ static void uct_ib_md_close(uct_md_h uct_md)
     if (md->rcache != NULL) {
         ucs_rcache_destroy(md->rcache);
     }
+    if (md->global_mem.atomic_mr != NULL) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: need to extract this to common function which would be called from error flow as well

File: src/uct/ib/base/ib_md.h [hunk 2]
```diff
@@ -76,6 +76,9 @@ typedef struct uct_ib_md {
     struct ibv_pd            *pd;       /**< IB memory domain */
     uct_ib_device_t          dev;       /**< IB device */
     uct_linear_growth_t      reg_cost;  /**< Memory registration cost */
+    int                      eth_pause; /**< Pause Frame on an Ethernet network */
+//    uct_ib_odp_config_t      odp;       /**< ODP configuration */
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h]: pls remove

File: test/gtest/uct/test_pd.cc [hunk 3]
```diff
@@ -6,11 +6,14 @@
 
 extern "C" {
 #include <uct/api/uct.h>
+#include <uct/ib/base/ib_md.h>
```
Comment (yosefe) [inline test/gtest/uct/test_pd.cc]: this unit test is generic and not ib specific, need to move to test_ib.cc

File: test/gtest/uct/test_pd.cc [hunk 4]
```diff
@@ -83,20 +91,101 @@ std::vector<std::string> test_pd::enum_mds(const std::string& mdc_name) {
...
+
+    support_odp = 0;
+#if HAVE_STRUCT_IBV_EXP_DEVICE_ATTR_ODP_CAPS
+    if (reg_method == "odp") {
+        orig_dev_list = dev_list = ibv_get_device_list(&num_devices);
+        if (!dev_list) {
+            UCS_TEST_ABORT("Failed to get the device list.");
+        }
+
+        while (*dev_list) {
```
Comment (yosefe) [inline test/gtest/uct/test_pd.cc]: please query for odp support from the IB MD structure (for example from uct_ib_device_t)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4091/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4092/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4093/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4094/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4095/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4096/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4099/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:mlx:reboot

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -337,8 +337,10 @@ enum uct_md_mem_flags {
                                                 mapping may be deferred until
                                                 it is accessed by the CPU or a
                                                 transport. */
-    UCT_MD_MEM_FLAG_FIXED    = UCS_BIT(1)  /**< Place the mapping at exactly
+    UCT_MD_MEM_FLAG_FIXED    = UCS_BIT(1), /**< Place the mapping at exactly
                                                 defined address */
+    UCT_MD_MEM_ODP_DISABLE   = UCS_BIT(2)  /**< Memory can't be used with
```
Comment (shamisp) [inline src/uct/api/uct.h]: On demand paging ?

**Pull request closed without merging**